// dimensions/blacklistcheckclient.ts
import ClientArgs from 'dimensions/clientargs';
import RawPacket from 'dimensions/packets/rawpacket';
import { getPacketsFromBuffer, BuffersPackets } from 'dimensions/utils';
import { getProperIP } from 'dimensions/utils';
import Blacklist from 'dimensions/blacklist';
import * as PlayerSlotSet from 'terraria-packet/src/packet/Packet_PlayerSlotSet.gen';
import * as Parser from 'terraria-packet/src/Parser.gen';

enum ClientState {
    StartOfConnection,
    AssignedClientId,
    SentPlayerInfo,
    SentUuid,
}

interface BlacklistCheckClientArgs {
    blacklist: Blacklist,
    clientArgs: ClientArgs,
    clientAcceptedCb: (packetsReceived: RawPacket[]) => void,
    clientBlacklistedCb: () => void
    errorCheckingBlacklistCb: (packetsReceived: RawPacket[], e: Error) => void
    packetErrorCheckingBlacklistCb: (e: Error) => void
}

class BlacklistCheckClient {
    private state: ClientState = ClientState.StartOfConnection;
    private name: string | undefined;
    private bufferPacket: Buffer = Buffer.alloc(0);
    private packetsReceived: RawPacket[] = [];

    constructor(private args: BlacklistCheckClientArgs) {
        args.clientArgs.socket.on('data', this.handleData.bind(this));
        args.clientArgs.socket.on('error', this.handleError.bind(this));
        args.clientArgs.socket.on('timeout', this.handleTimeout.bind(this));
        args.clientArgs.socket.on('close', () => {
            this.dispose()
        })
        args.clientArgs.socket.write(
            PlayerSlotSet.toBuffer(0)
        );
        this.state = ClientState.AssignedClientId;
    }

    handleData(data: Buffer) {
        let bufferPacket = this.bufferPacket;
        let entireData = Buffer.concat([bufferPacket, data]);

        // Get the individual packets from the data
        let entireDataInfo: BuffersPackets = getPacketsFromBuffer(entireData);

        // Update Buffer Packet using the new incomplete packet (if any)
        this.bufferPacket = entireDataInfo.bufferPacket;

        const packets: RawPacket[] = entireDataInfo.packets;
        this.packetsReceived.push(...packets);

        for (const packet of packets) {
            this.handlePacket(packet);
        }
    }

    handlePacket(rawPacket: RawPacket) {
        const packet = Parser.parseLazy(rawPacket.data, false)
        if (packet == undefined) {
            return
        }

        switch (packet.TAG) {
            case "PlayerInfo":
                if (this.state !== ClientState.AssignedClientId) {
                    this.dispose()
                    this.args.packetErrorCheckingBlacklistCb(new Error("Client info packet received before assigning client ID"))
                    return
                }
                const playerInfo = packet._0.VAL();
                if (playerInfo === undefined) {
                    this.dispose()
                    this.args.packetErrorCheckingBlacklistCb(new Error("Client info packet could not be parsed"))
                    return
                }

                this.name = playerInfo.name;
                this.state = ClientState.SentPlayerInfo;
                break;
            case "ClientUuid":
                if (this.state !== ClientState.SentPlayerInfo) {
                    this.dispose()
                    this.args.packetErrorCheckingBlacklistCb(new Error("Client UUID packet received before player info"))
                    return
                }
                const clientUuid = packet._0.VAL();
                const ip = getProperIP(this.args.clientArgs.socket.remoteAddress);
                if (clientUuid === undefined) {
                    this.dispose()
                    this.args.packetErrorCheckingBlacklistCb(new Error("Client UUID packet could not be parsed"))
                    return
                }
                if (this.name === undefined) {
                    this.dispose()
                    this.args.packetErrorCheckingBlacklistCb(new Error("Client name missing"))
                    return
                }
                if (ip === undefined) {
                    this.dispose()
                    this.args.packetErrorCheckingBlacklistCb(new Error("Client IP could not be parsed"))
                    return
                }

                this.args.blacklist.checkInformation(this.name, ip, clientUuid.uuid).then((isBlacklisted) => {
                    this.dispose()
                    if (isBlacklisted) {
                        this.args.clientBlacklistedCb();
                    } else {
                        this.args.clientAcceptedCb(this.packetsReceived);
                    }
                }).catch((e) => {
                    this.dispose()
                    this.args.errorCheckingBlacklistCb(this.packetsReceived, e)
                })
                break;
        }
    }

    handleError(err: Error) {
        console.error("Error with client", err);
    }
    handleTimeout() {
        console.error("Client timed out");
    }

    dispose() {
        this.args.clientArgs.socket.removeAllListeners();
    }
}

export default BlacklistCheckClient;
