// dimensions/language.ts
export interface LanguageDefinition {
  englishName: string,
  name: string,
  isoCode: string,
  phrases: LanguagePhrases,
}

export interface LanguagePhrasesOverrides {
  nameAlreadyOnServer?: string,
  characterNameLengthOutOfRange?: string,
  areYouEvenConnected?: string,
  youAreAlreadyInthatDimension?: string,
  shiftingToDimension?: string,
  youNeedToWaitUntilConnected?: string,
  playerCount?: string,
  availableDimensions?: string,
  youEnteredTheVoid?: string,
  dimensionDisconnectedYou?: string,
  reason?: string,
  dimensionsCommandName?: string,
  specifyADimensionToTravel?: string,
  dimensionDropped?: string,
  blacklisted?: string,
  blacklistCheckError?: string,
  close?: string,
}

interface LanguagePhrases {
  nameAlreadyOnServer: string,
  characterNameLengthOutOfRange: string,
  areYouEvenConnected: string,
  youAreAlreadyInthatDimension: string,
  shiftingToDimension: string,
  youNeedToWaitUntilConnected: string,
  playerCount: string,
  availableDimensions: string,
  youEnteredTheVoid: string,
  dimensionDisconnectedYou: string,
  reason: string,
  dimensionsCommandName: string,
  specifyADimensionToTravel: string,
  dimensionDropped: string,
  blacklisted: string,
  blacklistCheckError: string,
  close: string,
}

export const english: LanguageDefinition = {
  englishName: "English",
  name: "English",
  isoCode: "en",
  phrases: {
    nameAlreadyOnServer: "Someone called ${name} is already on the server.",
    characterNameLengthOutOfRange: "A character name must be between (inclusive) 2 to 20 characters long.",
    areYouEvenConnected: "Are you even connected?",
    youAreAlreadyInthatDimension: "You are already in that Dimension.",
    shiftingToDimension: "Shifting to the ${name} Dimension",
    youNeedToWaitUntilConnected: "You need to wait until you have fully connected to your current Dimension.",
    playerCount: "There are ${total} players across all Dimensions.",
    availableDimensions: "Available Dimensions: ",
    youEnteredTheVoid: "You have entered the Void. You will soon disappear.",
    dimensionDisconnectedYou: "The dimension you were in disconnected you",
    reason: "Reason: ",
    dimensionsCommandName: "dimensions",
    specifyADimensionToTravel: "Specify a [c/FF00CC:Dimension] to travel to:",
    dimensionDropped: "The dimension you were in dropped the connection.",
    blacklisted: "You are blacklisted from the server.",
    blacklistCheckError: "There was an error checking if you are blacklisted.",
    // This is just on the assumption that in production you are not permanently
    // closing the server, but rather restarting it.
    close: "The server is being restarted. Please rejoin.",
  },
};

export const chinese: LanguageDefinition = {
  englishName: "Chinese",
  name: "汉语",
  isoCode: "zn",
  phrases: {
    nameAlreadyOnServer: "${name} 已经进入了服务器",
    characterNameLengthOutOfRange: "昵称长度必须在2至20个字符之间",
    areYouEvenConnected: "请确保你已经连接",
    youAreAlreadyInthatDimension: "你已经进入了那个世界",
    shiftingToDimension: "正在传送到 ${name}",
    youNeedToWaitUntilConnected: "请等待传送",
    playerCount: "当前共有 ${total} 名玩家在线",
    availableDimensions: "可传送的世界: ",
    youEnteredTheVoid: "你已断线，即将被踢出世界",
    dimensionDisconnectedYou: "你已与当前世界断开连接",
    reason: "原因:: ",
    dimensionsCommandName: "世界",
    specifyADimensionToTravel: "请选择需要传送的[c/FF00CC:世界]：",
    dimensionDropped: "你所在的世界已与您断开连接",
    blacklisted: "您已被服务器列入黑名单。",
    blacklistCheckError: "检查您是否被列入黑名单时出现错误。",
    // TODO: This is google translated, should be reviewed by native speaker
    close: "服务器正在重新启动。 请重新加入。",
  },
};
