// dimensions/extension/index.ts
import Packet from 'dimensions/packet';
import Client from 'dimensions/client';
import TerrariaServer from 'dimensions/terrariaserver';
import ClientPacketHandler from 'dimensions/extension/clientpackethandler';
import TerrariaServerPacketHandler from 'dimensions/extension/terrariaserverpackethandler';

export interface PacketHandler {
    clientHandler?: ClientPacketHandler;
    serverHandler?: TerrariaServerPacketHandler;
}

type ClientErrorHandler = (client: Client, error: Error) => boolean;
type ServerErrorHandler = (server: TerrariaServer, error: Error) => boolean;
type ClientDisconnectHandler = (client: Client) => boolean;
type ServerDisconnectHandler = (server: TerrariaServer) => boolean;

export interface Extension {
    name: string;
    version: string;
    author: string;
    reloadable: boolean;
    reloadName?: string;

    // Reloading
    reload?: (require: any) => void;

    // Packet Handling
    priorPacketHandlers: PacketHandler,
    postPacketHandlers: PacketHandler

    // Error Handling
    clientErrorHandler?: ClientErrorHandler,
    serverErrorHandler?: ServerErrorHandler

    // Socket close Handling
    clientDisconnectPreHandler?: ClientDisconnectHandler,
    clientDisconnectHandler?: ClientDisconnectHandler,
    serverDisconnectPreHandler?: ServerDisconnectHandler
    serverDisconnectHandler?: ServerDisconnectHandler
}

export default Extension;
