// dimensions/packets/hexreader.ts
import Color from "dimensions/color";
import NetworkText from "dimensions/packets/networktext";
import { hex2a } from "dimensions/utils";
import * as utf8 from "utf8";

/**
 * Reads datatypes from a hex string
 */
class BufferReader {
    protected _data: Buffer;
    protected _head: number = 0;

    constructor(data: Buffer) {
        this._data = data;
    }

    /**
     * Gets the data still stored by the reader
     *
     * @return The hex data still stored by the reader
     */
    public get data() {
        return this._data;
    }

    public set buffer(data: Buffer) {
        this._data = data;
        this._head = 0;
    }

    public set head(head: number) {
        this._head = head;
    }

    /**
     * Removes a byte from the data
     *
     * @return The byte removed
     */
    public readByte(): number {
        const byte = this._data.readUInt8(this._head);
        this._head += 1;

        return byte;
    }

    /**
     * Removes a certain number of bytes from the data
     *
     * @return The bytes removed
     */
    public readBytes(amount: number): string {
        let bytes = "";
        for (let i = 0; i < amount; i++) {
            bytes += this.readByte();
        }

        return bytes;
    }

    /**
     * Removes three bytes from the data and puts them into a color object
     *
     * @return The color object {R, G, B}
     */
    public readColor(): Color {
        const color = {
            R: this.readByte(),
            G: this.readByte(),
            B: this.readByte()
        };

        return color;
    }

    /**
     * Removes a byte and converts it into an sbyte
     *
     * @return The sbyte removed
     */
    public readSByte(): number {
        const byte = this._data.readInt8(this._head);
        this._head += 1;

        return byte;
    }

    /**
     * Removes 2 bytes from the data, converting it to a signed int16
     * as an int16.
     *
     * @return An int16
     */
    public readInt16(): number {
        const int16 = this._data.readInt16LE(this._head);
        this._head += 2;

        return int16;
    }

    /**
     * Removes 2 bytes from the data and converts it to an unsigned int16
     *
     * @return A uint16
     */
    public readUInt16(): number {
        const uint16 = this._data.readUInt16LE(this._head);
        this._head += 2;

        return uint16;
    }

    /**
     * Removes 4 bytes from the data, converting it into a signed int32
     *
     * Perf: https://jsperf.com/conv-dimen-1
     * @return The signed int32 removed from the data
     */
    public readInt32(): number {
        const int32 = this._data.readInt32LE(this._head);
        this._head += 4;

        return int32;
    }

    /**
     * Removes 4 bytes from the data, converting it into an unsigned int32
     *
     * @return The unsigned int32 removed from the data
     */
    public readUInt32(): number {
        const uint32 = this._data.readUInt32LE(this._head);
        this._head += 4;

        return uint32;
    }

    /**
     * Removes 8 bytes from the data, converting it into an unsigned int64
     *
     * @return The unsigned int64 removed from the data
     */
    public readUInt64(): number {
        const str = this._data.toString("hex");
        const low = this.readUInt32();
        const high = this.readUInt32();

        return low + high * 4294967296.0;
    }

    /**
     * Removes 4 bytes from the data converting it into a single
     *
     * @return The single removed from the data
     */
    public readSingle(): number {
        const single = this._data.readFloatLE(this._head);
        this._head += 8;

        return single;
    }

    /**
     * Removes a certain number of bytes for the string length, and then further
     * bytes depending on the length, converting it into a string of characters.
     *
     * @return The string removed from the data
     */
    public readString(): string {
        // Read string length
        const firstByte: number = this.readByte();
        let strLength: number = firstByte;
        if (firstByte >= 128) {
            const secondByte: number = this.readByte();
            strLength = firstByte + secondByte << 8;
        }

        // Read string content using length
        const strContent: string = utf8.decode(hex2a(this._data.slice(this._head, this._head + strLength).toString("hex")));
        this._head += strLength;
        return strContent;
    }

    /**
     * Removes the mode byte and the string bytes
     *
     * @return The text
     */
    public readNetworkText(): NetworkText {
        const mode = this.readByte();
        const text = this.readString();

        return new NetworkText(mode, text);
    }
}

export default BufferReader;
