// dimensions/player.ts
import Item from 'dimensions/item';
import Client from 'dimensions/client';
import PacketWriter from 'dimensions/packets/packetwriter';
import PacketTypes from 'dimensions/packettypes';
import Color from 'dimensions/color';

class Player {
	public id: number;
	public name: string;
	public inventory: Item[];
	public life: number;
	public mana: number;
	public allowedNameChange: boolean;
	public allowedCharacterChange: boolean;
	public allowedLifeChange: boolean;
	public allowedManaChange: boolean;
	public position: { x: number, y: number } = {
		x: 0,
		y: 0
	};

	// Visuals
	public skinVariant: number;
	public hair: number;
	public hairDye: number;
	public hideVisuals: number;
	public hideVisuals2: number;
	public hideMisc: number;
	public hairColor: Color;
	public skinColor: Color;
	public eyeColor: Color;
	public shirtColor: Color;
	public underShirtColor: Color;
	public pantsColor: Color;
	public shoeColor: Color;
	private client: Client | null;

	// Soft/Med/Hard core
	public difficulty: number;

	constructor(client: Client | null) {
		this.client = client;
		this.id = 0;
		this.name = "";
		this.life = 100;
		this.mana = 20;
		this.allowedNameChange = false;
		this.allowedCharacterChange = false;

		// Inventory of Client - only used for SSC -> to Non-SSC switching
		this.inventory = [];
	}

	public setItem(item: Item): void {
		if (this.client === null) {
			return;
		}

		let playerInventorySlot: string = (new PacketWriter())
			.setType(PacketTypes.PlayerInventorySlot)
			.packByte(this.id)
			.packByte(item.slot)
			.packInt16(item.stack)
			.packByte(item.prefix)
			.packInt16(item.netID)
			.data;

		this.client.socket.write(new Buffer(playerInventorySlot, 'hex'));
  	}

	  public setLife(life: number): void {
		if (this.client === null) {
			return;
		}

		let playerLife: string = (new PacketWriter())
			.setType(PacketTypes.PlayerHP)
			.packByte(this.id)
			.packInt16(this.life)
			.packInt16(this.life)
			.data;

		this.client.socket.write(new Buffer(playerLife, 'hex'));
	}

	public setMana(mana: number): void {
		if (this.client === null) {
			return;
		}

		let playerMana: string = (new PacketWriter())
			.setType(PacketTypes.PlayerMana)
			.packByte(this.id)
			.packInt16(this.mana)
			.packInt16(this.mana)
			.data;

		this.client.socket.write(new Buffer(playerMana, 'hex'));
	}

	public setVisuals(): void {
		if (this.client === null) {
			return;
		}

		let playerInfo: string = (new PacketWriter())
			.setType(PacketTypes.PlayerInfo)
			.packByte(this.id)
			.packByte(this.skinVariant)
			.packByte(this.hair)
			.packString(this.name)
			.packByte(this.hairDye)
			.packByte(this.hideVisuals)
			.packByte(this.hideVisuals2)
			.packByte(this.hideMisc)
			.packColor(this.hairColor)
			.packColor(this.skinColor)
			.packColor(this.eyeColor)
			.packColor(this.shirtColor)
			.packColor(this.underShirtColor)
			.packColor(this.pantsColor)
			.packColor(this.shoeColor)
			.packByte(this.difficulty)
			.data;

		this.client.socket.write(new Buffer(playerInfo, 'hex'));
	}
}

export default Player;
